toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt

```diff
@@ -0,0 +1,358 @@
+                        element,
+                        UtilityAssociationsElementState(element, coroutineScope)
+                    )
+                }
+
+                else -> {
+                    // TODO remove for release
+                    println("encountered element of type ${element::class.java}")
+                }
+            }
+        }
+
+        return states
+    }
+}
+
+/**
+ * A structure that holds the [Popup] and its associated [PopupElementStateCollection].
+ *
+ * This class is also [Stable] and enables composition optimizations.
+ *
+ * @param popup the [Popup] to create the state for.
+ * @param stateCollection the [PopupElementStateCollection] created for the [Popup].
+ */
+@Stable
+internal data class PopupStateData(
+    val popup: Popup,
+) {
+    internal lateinit var stateCollection: PopupElementStateCollection
+        private set
+    /**
+     * Indicates if the evaluateExpression function for the [popup] has been run.
+     */
+    internal var initialEvaluation : MutableState<Boolean> = mutableStateOf(false)
+        private set
+
+    /**
+     * Indicates if the expressions for the [popup] are currently being evaluated.
+     */
+    internal var isEvaluatingExpressions: MutableState<Boolean> = mutableStateOf(false)
+        private set
+
+    internal fun setStates(stateCollection: PopupElementStateCollection) {
+        this.stateCollection = stateCollection
+    }
+
+    /**
+     * Evaluates the expressions for the [popup] and returns all the attachments. After a successful
+     * evaluation, the [initialEvaluation] is set to true. While this function is running, the
+     * [isEvaluatingExpressions] will be true.
+     */
+    internal suspend fun evaluateExpressionsAndGetAttachments() : List<PopupAttachment>  {
+        try {
+            isEvaluatingExpressions.value = true
+            val attachments = mutableListOf<PopupAttachment>()
+            popup.evaluateExpressions().onSuccess {
+                val element = popup.evaluatedElements
+                    .filterIsInstance<AttachmentsPopupElement>()
+                    .firstOrNull()
+
+                element?.fetchAttachments()?.onSuccess {
+                    attachments.addAll(element.attachments)
+                }
+                // Set the initial evaluation to true after the first successful evaluation.
+                initialEvaluation.value = true
+            }
+            return attachments
+        } finally {
+            isEvaluatingExpressions.value = false
+        }
+    }
+}
+
+internal fun ArcGISFeature.toPopup(): Popup {
+    val popupDefinition = when {
+        this.featureTable?.popupDefinition != null -> {
+            this.featureTable?.popupDefinition
+        }
+
+        this.getFeatureSubtype() != null && this.featureTable is ArcGISFeatureTable -> {
+            (this.featureTable as ArcGISFeatureTable).subtypeSubtables
+                .firstOrNull { it.name == this.getFeatureSubtype()?.name }
```

Comment:
```suggestion
                .firstOrNull { it.subtype == this.getFeatureSubtype()}
```
The `FeatureSubtype` can be checked directly here.

Reply:
updated